src/render.rs

```diff
@@ -523,6 +523,38 @@ impl<'a, ComponentId: Clone + Debug + Eq + Hash> Viewport<'a, ComponentId> {
         span_rect
     }
 
+    /// Draw a [`Line`] directly to the screen at `(x, y)` location.
+    /// If the contents would overflow the current mask,
+    /// then it is truncated with an ellipsis.
+    pub fn draw_line(&mut self, x: isize, y: isize, line: &Line) -> Rect {
+        let span_rect = Rect {
+            x,
+            y,
+            width: line.width(),
+            height: 1,
+        };
+        self.current_trace_mut().merge_rect(span_rect);
+
+        let draw_rect = self.rect.intersect(span_rect);
+
+        let draw_rect = match self.mask {
+            Some(mask) => mask.apply(draw_rect),
+            None => draw_rect,
+        };
+        if !draw_rect.is_empty() {
+            let buf_rect = self.translate_rect(draw_rect);
+            line.render(buf_rect, self.buf);
+
+            let mask_rect = self.mask_rect();
+
+            if span_rect.end_x() > mask_rect.end_x() {
+                self.draw_span(mask_rect.end_x() - 1, span_rect.y, &Span::raw("…"));
+            }
+        }
+
+        draw_rect
+    }
```

Comment:
note: I'm not sure this was the "right" way to go about implementing rendering for `Line`, or if the ellipsis should even be done in _this_ function, so don't take anything here too seriously.

Reply:
I think we should keep the ellipsis behavior in one function. Can you change `draw_text` to accept a `Line` instead of a `Span`? You could have it call this function, or you could inline this function into `draw_text` if appropriate. (I think you can?)

It looks like practically every call site passes a reference to a temporary `Span`, so to make it easier, you could give the function this signature:

```rs
    pub fn draw_text<'b>(&mut self, x: isize, y: isize, line: impl Into<Line<'b>>) -> Rect { ... }
```

and then you can just change `draw_text(x, y, &Span::styled(...))` to `draw_text(x, y, Span::styled(...))`, etc., in the appropriate places, rather than explicitly create real `Line` instances everywhere.

Reply:
Check out [`0b2a273` (#37)](https://github.com/arxanas/scm-record/pull/37/commits/0b2a27361180a74c2926c873a1df42b808b414f3)